workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/flows/RedeemToken.kt

```diff
@@ -1,85 +1,172 @@
+import net.corda.core.flows.InitiatedBy
+import net.corda.core.flows.InitiatingFlow
+import net.corda.core.flows.ReceiveFinalityFlow
+import net.corda.core.flows.ReceiveStateAndRefFlow
+import net.corda.core.flows.SendStateAndRefFlow
+import net.corda.core.flows.SignTransactionFlow
+import net.corda.core.flows.StartableByRPC
+import net.corda.core.identity.Party
+import net.corda.core.node.StatesToRecord
+import net.corda.core.node.services.IdentityService
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import net.corda.core.utilities.ProgressTracker
+import net.corda.core.utilities.unwrap
+
+object RedeemToken {
+
+    @CordaSerializable
+    data class TokenRedeemNotification<T : TokenType>(val anonymous: Boolean, val amount: Amount<T>?)
+
+    // TODO later split it into non fungible or fungible
+    // Called on owner side.
+    @InitiatingFlow
+    @StartableByRPC
+    class InitiateRedeem<T : TokenType>(
+            val ownedToken: T,
+            val issuer: Party,
+            val amount: Amount<T>? = null,
+            val anonymous: Boolean = true
+    ) : FlowLogic<SignedTransaction>() {
+        companion object {
+            object REDEEM_NOTIFICATION : ProgressTracker.Step("Sending redeem notification to parties.")
+            object CONF_ID : ProgressTracker.Step("Requesting confidential identity.")
+            object SELECTING_STATES : ProgressTracker.Step("Selecting states to redeem.")
+            object SEND_STATE_REF : ProgressTracker.Step("Sending states to the issuer for redeeming.")
+            object SIGNING_TX : ProgressTracker.Step("Signing transaction")
+            object FINALISING_TX : ProgressTracker.Step("Finalising transaction")
+
+            fun tracker() = ProgressTracker(REDEEM_NOTIFICATION, CONF_ID, SELECTING_STATES, SEND_STATE_REF, SIGNING_TX, FINALISING_TX)
+        }
+
+        override val progressTracker: ProgressTracker = tracker()
+
+        @Suspendable
+        override fun call(): SignedTransaction {
+            val issuerSession = initiateFlow(issuer)
+
+            progressTracker.currentStep = REDEEM_NOTIFICATION
+            // Notify the recipient that we'll be sending them tokens for redeeming and advise them of anything they must do, e.g.
+            // request a confidential identity.
+            issuerSession.send(TokenRedeemNotification(anonymous = anonymous, amount = amount))
+
+            // Generate and send over a new confidential identity, if necessary.
+            if (anonymous) {
+                progressTracker.currentStep = CONF_ID
+                subFlow(RequestConfidentialIdentity.Responder(issuerSession))
+            }
+
+            progressTracker.currentStep = SELECTING_STATES
+            val exitStateAndRefs = if (amount == null) {
+                // TODO We should be able to figure out the issuer based on the issuer of this token.
```

Comment:
yeah if we pass in an `IssuedTokenType` instead of a token type, you can do this.

Reply:
This comments refers more to the fact that all the flows in Token sdk take amount + issuer for fungible and token for non-fungible. This will go after cleaning up of flow API.